### ISSUE TITLE
The system cannot find the path specified. Windows Error

### ISSUE THREAD

**Issue by laggingreflex** [nathanboktae/mocha-casperjs#25]
> The system cannot find the path specified.

Getting this error on Windows(8x64) when running `mocha-casperjs.cmd`

---

I notice that `mocha-casperjs.cmd` is looking for `/bin/sh.exe`

<pre>
@IF EXIST "%~dp0\/bin/sh.exe" (
  "%~dp0\/bin/sh.exe"  "%~dp0\node_modules\mocha-casperjs\bin\mocha-casperjs" %*
) ELSE (
  /bin/sh  "%~dp0\node_modules\mocha-casperjs\bin\mocha-casperjs" %*
)
</pre>

I don't know enough batch but my `sh.exe` is
![](http://i.imgur.com/ek2xfCt.png)
And even on replacing the path to above manually it gives this error

> <pre>
> mocha-casperjs
> cygwin warning:
>   MS-DOS style path detected: C:\...\node_modules\mocha-casperjs\bin\mocha-casperjs
>   Preferred POSIX equivalent is: /cygdrive/c/Users/x/AppData/Roaming/npm/node_modules/mocha-casperjs/bin/mocha-casperjs
>   CYGWIN environment variable option "nodosfilewarning" turns off this warning.
>   Consult the user's guide for more details about POSIX paths:
>     http://cygwin.com/cygwin-ug-net/using.html#using-pathnames
> C:\...\node_modules\mocha-casperjs\bin\mocha-casperjs: line 3: dirname: command not found
> C:\...\node_modules\mocha-casperjs\bin\mocha-casperjs: line 4: readlink: command not found
> C:\...\node_modules\mocha-casperjs\bin\mocha-casperjs: line 9: /../../casperjs/bin/casperjs: No such file or directory
> </pre>
> 

The `mocha.cmd` just looks for `node.exe`

<pre>
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\node_modules\mocha\bin\mocha" %*
) ELSE (
  node  "%~dp0\node_modules\mocha\bin\mocha" %*
)
</pre>

**Comment by eugeniol** [comment]
I have the same issue on windows, I have cygwin installed but i'm not running npm on cygwin
Why does it look for sh.exe?

**Comment by Scthe** [comment]
Can confirm the problem, temporary fix:

1. open **node_modules\grunt-mocha-casperjs\node_modules\.bin\mocha-casperjs.cmd**
2. replace line 4 with: `"%~dp0\..\mocha-casperjs\bin\mocha-casperjs" %*`
3. open **node_modules\grunt-mocha-casperjs\node_modules\mocha-casperjs\bin\mocha-casperjs.bat**
4. replace last line with: `%MOCHA_CASPER_PATH%.\..\..\casperjs\bin\casperjs.exe %MOCHA_CASPER_PATH%\cli.js --mocha-casperjs-path=%MOCHA_CASPER_PATH%.. %*`

ofc. This *fix* may be in some way specific to my configuration..

**Comment by nathanboktae** [comment]
I'm confused - that's exactly what's in `mocha-casperjs.bat` [right now](https://github.com/nathanboktae/mocha-casperjs/blob/a25a8b6e0540ee7323e5cba7ae6961e3a246149b/bin/mocha-casperjs.bat#L36)

**Comment by Scthe** [comment]
Yes, You are right, I have not checked that. I'm sorry for the confusion. 

As for the **node_modules\.bin\mocha-casperjs.cmd** mentioned before if it is npm that creates and owns .cmd I'm ok with Your explanation ( message from Jun 24, 2014). Still, following steps 1&2 from the instruction above solved my problem and if someone ever had similar issues I hope it will help too. ( I've updated my node & npm version and the problem persisted).

Again, sorry for confusion and thanks for fast reply. This project is awesome.

PS. Although my previous message indicates that I'm using Your project through grunt-mocha-casperjs I've tested through mocha-casperjs too.

**Comment by pietrofxq** [comment]
I'm experiencing a similar issue. This is my output when I run casper-mochajs command:

    dirname: too many arguments
    try 'dirname --help' for more information.
    /c/users/Pietro Coelho/AppData/Roaming/npm/node_modules/mocha-casperjs/bin/mocha-casperjs: line 31: readlink: command not found
    /c/Users/Pietro Coelho/AppData/Roaming/npm/node_modules/mocha-casperjs/bin/mocha-casperjs: line 36: /../../casperjs/bin/casperjs: No such file or directory

I have phantomjs and casperjs available from the command line.

I'm using Windows 8 64 bit with git bash

**Comment by cyberspeeyush** [comment]
I am getting the same error.. Any help will be appreciated.

**Comment by webtobesocial** [comment]
Hi everyone,

as I tried to use `mocha-casperjs`, I got an error.

```bash
$ mocha-casperjs --v
C:\Users\foobar\AppData\Roaming\npm/node_modules/mocha-casperjs/bin/mocha-casperjs: line 39: C:\Users: command not found
```

**Enviroment**:
- windows 7
- cygwin/babun: 1.2.0
- zsh: 5.0.6
- node: 0.12.7
- npm: 2.11.3

I tried to the fix from @Scthe but that didn't work. Any help will be appreciated.

**Comment by dredtrakauf** [comment]
Hi,
had the same issue,
I just changed my mocha-casperjs.bat file to

@IF EXIST "%~dp0\/bin/sh.exe" (
  "%~dp0\/bin/sh.exe"  "%~dp0\node_modules\mocha-casperjs\bin\mocha-casperjs" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  "%~dp0node_modules\mocha-casperjs\bin\mocha-casperjs" %*
)
and it worked!

**Comment by basherr** [comment]
I've even installed it globally but still getting the same error.

**Comment by jhdevuk** [comment]
@nathanboktae I'm also getting this error. Any idea on fixes? Thanks

**Comment by nathanboktae** [comment]
I highly recommend moving on to nightmarejs now. See the depreciation notice I posted. Otherwise there's instructions for windows users in the readme and help here.

**Comment by jahilldev** [comment]
Nice, +1 for the great link. Haven't come across nightmarejs before. Thanks

**Comment by basherr** [comment]
@nathanboktae Better suggest `Chrome Headless` rather than `nightmare js`. I've already used `nightmare api` and encountered few issues. The repo owner is no more interested to maintain it any more.

**Comment by nathanboktae** [comment]
> I've already used nightmare api and encountered few issues. The repo owner is no more interested to maintain it any more.

I think you're talking about something else. [nightmarejs](https://github.com/segmentio/nightmare) has 15,605 stars as of writing this, released 3.0.1 on March, and is maintained not by a person but a company, Segment.io, which has [raised 109million in funding](https://www.crunchbase.com/organization/segment-io).

I really like headless chrome myself too, and I recommended [mocha-chrome](https://github.com/shellscape/mocha-chrome/issues) for `mocha-phantomjs` users. I have also used [SauceLabs](https://saucelabs.com/open-source) before when I needed real browsers with good success.
